### DISCUSSION TITLE
Some questions about two-phase flow and RELAP7

### DISCUSSION THREAD

**Discussion by SomnusYu** [idaholab/moose#26388] [Q&A General] [answered]
Hi @GiudGiud and @joshuahansel,

These days I would like to develop a two-phase six-equation model using MOOSE, which seems easier than the seven-equation model that was used in RELAP7. Which of these two models is more suitable or easier to develop in MOOSE?
By the way, I would like to know if the RELAP7 can calculate some accident conditions, like LOCA, RIA, LOFA..... Because I can not find any reference about this. If yes, could you please kindly show me some references?

Best regards.

**Comment by GiudGiud** [comment]
@licharlot will know which tech reports have accidental transients with RELAP7

**Comment by joshuahansel** [comment]
To my knowledge, I don't think either model is easier to develop than the other. RELAP-7 is capable of simulating accident conditions like the ones you listed. The only question is whether it will be robust enough to finish your transient through any phase disappearance (zero volume fraction). As far as why RELAP5 has been more successful in this area (I'm assuming this is the case anyway), I don't think it's necessarily 6-equation vs. 7-equation - I think it's just a better phase disappearance treatment. I think that both 6-equation and 7-equation models must treat phase disappearance in some way. I would say it may be not just a function of the flow model but also of the fluid property calculations robustness.

As for references, I would just see what comes up when you google "RELAP-7". I don't know that it's worth the effort of reading everything - just know that some accidents may succeed and some may fail due to phase disappearance robustness issues - it all depends on your problem.

**Reply by SomnusYu** [reply]
Thanks for you reply! Thank you very much. By the way, I also have other questions (I developed the six-equation model):
- what the variables of six-equation model should be? [α, p, ul, ug, Tl, Tg] or [(αρ)l, (αρ)g, (αρu)l, (αρu)g, (αρE)l, (αρE)g]? I know the former form is for incompressiable flow, and the later is for compressiable flow. Two-phase flow sometime is incompressiable(α=0), but sometimes it is compressiable(α=1). Which one should I choose?
- If α=0 or α=1, the 6-equation would become to 3-equation, which could cause non-convergence for NAN. Now, I just constrain the αmin = 1.0e-4 and αmax = 0.9999 to always use 6-equation. Is it correct?
- which fluid properties does RELAP7 use? Water97FluidProperty? Now, I use IAPWS-IF97. But it is a problem when it is single-phase. For example, when it is liquid (α <= αmin), should the vapour properties be liquid? If yes, the properties will change rapidly causing divergence. Now, I also constrain the properties using the variable p and Tsat+0.0001 to make sure they are always vapour properties.
- For the convection terms in 6-equation, I use roe-type numerical scheme and DG kernel, just like THM. But it just calculate convection term. The momentum equations of six-equation model have the terms in red circle. How to solve this term? Now, I just put it in Material using (αi+1 - αi)/xi+1 - xi. Is it correct? Is there any better way?
![image](https://github.com/idaholab/moose/assets/118275988/a56eb34e-22f0-488d-808e-53110c97fc78)


Thank you very much! I'm so sorry I have so many questions.....

**Reply by joshuahansel** [reply, marked as answer]
- The variables can be either of those sets. We chose the latter set for the 7-equation model, but I've been wondering if we should additionally implement the former. In terms of the formulation, it does not matter, but I do believe it may have an impact on the numerics, perhaps the conditioning of your Jacobian dF/dU. You say alpha=0 and alpha=1 are related to compressibility? I don't understand that - it just determines which phase is present.
- You can try that strategy. We certainly have, and it was not a silver bullet. Maybe it'll work out for you.
- Typically we used the 1995 water formulation in RELAP-7. We could use the 1997 formulation if all of the necessary interfaces were implemented. And your questions indicate the tricky question - what happens to the vanished phase? A good question that we don't quite have the answer to. We have some relaxation terms that would keep that phase around saturation conditions of the other phase. That may work.
- The 7-equation model has similar terms but uses the interfacial pressure instead. That term becomes a non-conservative contribution to our inter-cell flux vector, so it is treated in a similar manner as the advective terms. Otherwise you could try it as a volumetric term if you're able to produce some approximation like you've suggested within the cell volume.

In general, and I cannot speak for everyone on the MOOSE team, but we don't have all of the answers to the questions of what's best for a two-phase flow formulation. We only know what we attempted, and we did not have the time and funding to produce a formulation that was entirely satisfactory, so I'd suggest you do some research into the methodology from a literature review. What we can definitely help you with is the implementation in MOOSE. The rest of the help we give you may be questionable.

**Reply by lindsayad** [reply]
Tagging @joe61vette who has quite a bit of experience with phase disappearance and attempts to model it with MOOSE (mostly unsuccessful I fear)

**Reply by lindsayad** [reply]
Just a note that the `navier_stokes` module has both incompressible and compressible formulations for single phase flow. I think that in the future if/when we add an incompressible formulation in the `thermal_hydraulics` module, we will leverage a lot of what we have in `navier_stokes`

**Reply by SomnusYu** [reply]
hi @joshuahansel,

Thanks for your detailed reply.

- what I said α is αg. αg = 0 means it is liquid, since the mass, momentum and energy equation on the left side is equal to 0 for vapour phase. In this case, it is incompressible. On the other hand, if αg = 1, it is pure vapour, which means it is compressible. As you said, the formulation has impact on the numerics, so which formulation is better for the two-phase six-equation model?
- Could you please kindly tell me the strategy you used? or could you give me some reference? I think my strategy is not very good.
- For the properties, I noticed RELAP7 also has the stiffened gas equation of state (SGEOS). Does it work? Although SGEOS is not accurate enough, I think it does not have the same problem, like IF-97.
-  when you treat this term, do you use the numerical scheme, like HLLC or something else? Additionally, If we treat this term in the similar manner, it is necessary to implement an additional boundary condition, what the boundary flux should be? (flux_advective + p*flux_alpha) * _normal * _test[_i][_qp]?

Thank you very much. These issues will not be encountered unless we are truly writing these codes. So it is hard for me to find these answers in literature. It needs the experiences that develop two-phase flow, which is exactly lacking for me. Your suggestion is really useful for me. Thank you again.

**Reply by joshuahansel** [reply]
- Ok
- Essentially this was the strategy we used, but we may have done it in a different way. We used a "remapper" such that the actual solution could go out of the range (0,1), but it would be uniquely mapped to numbers within (0,1). Yes, it keeps volume fraction in the range, but there are other issues.
- SGEOS works and is much more robust than other options, but even it is not immune to phase disappearance issues.
- Yes, you would have a corresponding flux on the boundary.

The best reference for details would be the RELAP-7 theory manual, which you can find by googling.

**Reply by SomnusYu** [reply]
OK, thanks for your patience！

**Reply by SomnusYu** [reply]
Hi @joshuahansel,

Could you please tell me the numerical method that RELAP7 used to deal with the p*∂α_g/∂x term? I searched through the RELAP7 theory manual, but I could not find it. Does it use the centered average of the left and right side fluxes (first-order)?

**Reply by SomnusYu** [reply]
I would like to have one more question. How to implement the p*∂α_g/∂x  term in ADDGkernel? I tried to implement this by adding this term in the DGkernel (according to ADNumericalFlux3EqnDGKernel.C) in the following way:

```
  ADReal re = 0.0;
  switch (type)  
  {
    case Moose::Element:
      if (_equation_index == EQ_LIQUID_MOMENTUM) 
        re = (flux_elem[_equation_index] + _press_interface[_qp]*flux_elem_nonc[_equation_index]) * _test[_i][_qp];
      else if (_equation_index == EQ_VAPOUR_MOMENTUM)
        re = (flux_elem[_equation_index] - _press_interface[_qp]*flux_elem_nonc[_equation_index]) * _test[_i][_qp];
      else
        re = flux_elem[_equation_index] * _test[_i][_qp];

      break;

    case Moose::Neighbor:
      if (_equation_index == EQ_LIQUID_MOMENTUM)
        re = -(flux_neig[_equation_index] + _press_interface_neig[_qp]*flux_neig_nonc[_equation_index]) * _test_neighbor[_i][_qp];
      else if (_equation_index ==EQ_VAPOUR_MOMENTUM)
        re = -(flux_neig[_equation_index] - _press_interface_neig[_qp]*flux_neig_nonc[_equation_index]) * _test_neighbor[_i][_qp];
      else
        re = -flux_neig[_equation_index] * _test_neighbor[_i][_qp];
      
      break;
  }

  return re;
```
![image](https://github.com/idaholab/moose/assets/118275988/2f1406d6-5e1d-4060-af7c-1062dd76c9e0)


In this way, the p*∂α_g/∂x also used the slope limiter. Does it correct?

**Reply by joshuahansel** [reply]
The `p*dalpha/dx` term was transformed into `d(p*alpha)/dx - alpha*dp/dx`. See the 2-phase flow spatial discretization section in the RELAP-7 theory manual (5.2.3 for me). The `p*alpha` in `d(p*alpha)/dx` is added to the conservative numerical flux, and the non-conservative term `alpha*dp/dx` term is integrated approximating `alpha` constant over the element.

**Reply by SomnusYu** [reply]
Thanks a lot. 

> the non-conservative term alpha*dp/dx term is integrated

I understand the way to deal with `alpha`, but I still do not fully understand the way to deal with the `dp/dx` term. Does this mean that the p is regarded as a variable and then the `coupledGradient` for variables is used to deal with `dp/dx`? However, the p is not the nonlinear variable, since the RELAP7 adopts the conservative variables.

**Reply by joshuahansel** [reply]
p does not need to be a solution variable. We do not use `coupledGradient`; we use a DGKernel just like with the conservative flux, but this time it's multiplied by alpha, so you end up with something like `alpha_i * p_{i+1/2}` on each side.

**Reply by SomnusYu** [reply]
Thanks! I got it.
